### PR TITLE
physnet max to dev

### PR DESCRIPTION
**Pull request by MaxH1996**

### CODE REVIEW

File: src/schnetpack/atomistic/electrostatic.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        
+        fac = self.kehalf*torch.gather(q, 0, idx_i)*torch.gather(q, 0, idx_j)
+        f = switch_function(rij, self.cuton, self.cutoff)
+        if self.lr_cutoff is None:
+            coulomb = 1/rij
+            damped  = 1/(rij**16 + self.cuton16)**(1/16)
+        else:
+            coulomb = torch.where(rij < self.lr_cutoff, 1.0/rij + rij/self.lr_cutoff**2 - 2.0/self.lr_cutoff, torch.zeros_like(rij))
+            damped  = 1/(rij**16 + self.cuton16)**(1/16) + (1-f)*self.cut_rconstant*rij - self.cut_constant
+        electrostatic_energy = q.new_zeros(N, dtype=torch.float).index_add_(0, idx_i, (fac*(f*damped + (1-f)*coulomb)).float())
```
[the remark this answers is not in the document]
Comment (ktschuett) [reply src/schnetpack/atomistic/electrostatic.py:60]: you can also replace index_add with schnetpack.nn.scatter here

[the remark this answers is not in the document]
Comment (ktschuett) [reply src/schnetpack/atomistic/electrostatic.py:60]: remove casting and set dtype=q.dtype

File: src/schnetpack/atomistic/zbl.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+'''
+class ZBLRepulsionEnergy(nn.Module):
+    def __init__(self, a0: float = 0.5291772105638411, ke: float =14.399645351950548):
+        super(ZBLRepulsionEnergy, self).__init__()
+        self.a0 = a0
+        self.ke = ke
+        self.kehalf = ke/2
+        self.register_parameter('_adiv', nn.Parameter(torch.tensor(1.)))
+        self.register_parameter('_apow', nn.Parameter(torch.tensor(1.)))
+        self.register_parameter('_c1', nn.Parameter(torch.tensor(1.)))
```
Comment (ktschuett) [inline src/schnetpack/atomistic/zbl.py:22]: replace by vectors `a, c` with shape (4,) and vectorize below in forward

File: src/schnetpack/atomistic/zbl.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        #normalize c coefficients (to get asymptotically correct behaviour for r -> 0)
+        csum = c1 + c2 + c3 + c4
+        c1 = c1/csum
+        c2 = c2/csum
+        c3 = c3/csum
+        c4 = c4/csum
+        #actual interactions
+        zizj = Zf[idx_i]*Zf[idx_j]
+        f = (c1*torch.exp(-a1*rij) + c2*torch.exp(-a2*rij) + c3*torch.exp(-a3*rij) + c4*torch.exp(-a4*rij))*cutoff_values
+        zbl_energy = Zf.new_zeros(N, dtype=torch.float).index_add_(0, idx_i, (self.kehalf*f*zizj/rij).float())
```
Comment (ktschuett) [inline src/schnetpack/atomistic/zbl.py:77]: schnetpack.nn.scatter

File: src/schnetpack/atomistic/zbl.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        self.register_parameter('_c2', nn.Parameter(torch.tensor(1.)))
+        self.register_parameter('_c3', nn.Parameter(torch.tensor(1.)))
+        self.register_parameter('_c4', nn.Parameter(torch.tensor(1.)))
+        self.register_parameter('_a1', nn.Parameter(torch.tensor(1.)))
+        self.register_parameter('_a2', nn.Parameter(torch.tensor(1.)))
+        self.register_parameter('_a3', nn.Parameter(torch.tensor(1.)))
+        self.register_parameter('_a4', nn.Parameter(torch.tensor(1.)))
+        #self.reset_parameters()
+        self.cutoff_fn = PhysNetCutOff(10.0)
+#     def reset_parameters(self):
```
Comment (ktschuett) [inline src/schnetpack/atomistic/zbl.py:32]: if not needed, delete

File: src/schnetpack/atomistic/D3.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+            kn=7.5,
+            wf=6.0,
+            Zmax=87,
+            Bohr=0.5291772105638411,
+            Hartree=27.211386024367243,
+            dtype=torch.float32):
+        super(D4DispersionEnergy, self).__init__()
+        assert Zmax <= 87
+        
+        #parameters
```
Comment (ktschuett) [inline src/schnetpack/atomistic/D3.py:52]: delete, if not needed

File: src/schnetpack/nn/blocks.py [hunk 6]
```diff
@@ -14,6 +14,7 @@ def build_mlp(
     n_hidden: Optional[Union[int, Sequence[int]]] = None,
     n_layers: int = 2,
     activation: Callable = F.silu,
+    weight_init: bool = False
```
Comment (ktschuett) [inline src/schnetpack/nn/blocks.py:17]: `last_init_zeros`?

File: src/schnetpack/nn/cutoff.py [hunk 7]
```diff
@@ -28,6 +28,23 @@ def cosine_cutoff(input: torch.Tensor, cutoff: torch.Tensor):
     return input_cut
 
 
+class PhysNetCutOff(nn.Module):
+    '''Cutoff Function from Physnet'''
```
Comment (ktschuett) [inline src/schnetpack/nn/cutoff.py:32]: add ref to paper

File: src/schnetpack/configs/data/custom.yaml [hunk 8]
```diff
@@ -6,6 +6,6 @@ batch_size: 10
 num_train: ???
 num_val: ???
 num_test: null
-num_workers: 4
+num_workers: 12
```
Comment (ktschuett) [inline src/schnetpack/configs/data/custom.yaml:9]: bit excessive. 4 should be a good default

File: src/schnetpack/configs/model/base_model.yaml [hunk 9]
```diff
@@ -6,6 +6,7 @@ _target_: schnetpack.atomistic.AtomisticModel
 optimizer_cls: torch.optim.Adam
 optimizer_args:
   lr: ${lr}
+  amsgrad: True
```
Comment (ktschuett) [inline src/schnetpack/configs/model/base_model.yaml:9]: leave default false

File: src/schnetpack/nn/radial.py [hunk 10]
```diff
@@ -77,7 +78,33 @@ def __init__(
 
     def forward(self, inputs: torch.Tensor):
         return gaussian_rbf(inputs, self.offsets, self.widths)
-
+    
+def softplus_inverse(x):
+    return x + np.log(-np.expm1(-x))
+
+class RBF_PhysNet(nn.Module):
```
Comment (ktschuett) [inline src/schnetpack/nn/radial.py:85]: more conistent with others: PhysNetRBF

File: src/schnetpack/properties.py [hunk 11]
```diff
@@ -44,7 +42,15 @@
...
 dipole_moment: Final[str] = "dipole_moment"
-polarizability: Final[str] = "polarizability"
+polarizability: Final[str] = "polarizability"  
+total_charge: Final[str] = "total_charge"
+
+    
+electrostatic: Final[str] = "electrostatic"
+zbl: Final[str] = "zbl"
+dispersion: Final[str] = "dispersion"
+physnet_aggregate: Final[str] = "physnet_aggregate"
```
Comment (ktschuett) [inline src/schnetpack/properties.py:55]: this should not be a constant here. you can internally call the atomwise energy `short_range` and then the aggregated energy just energy, or energy_U0 for QM9 (${property})

File: src/schnetpack/representation/support.py [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+        self.register_buffer("shift", torch.FloatTensor([shift]))
+
+    def forward(self, inputs: torch.Tensor):
+        
+        return snn.activations.shifted_softplus(inputs)  
+    
+def softplus_inverse(x):
+    return x + np.log(-np.expm1(-x))
+
+class RBF_PhysNet(nn.Module):
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:209]: duplicated

File: src/schnetpack/representation/support.py [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+        widths = [softplus_inverse((0.5/((1.0-np.exp(-cutoff))/n_rbf))**2)]*n_rbf
+        self.widths = snn.activations.shifted_softplus(torch.Tensor(widths))
+    def forward(self, r_ij):
+        r_ij = r_ij.unsqueeze(-1)
+        g_ij = torch.exp(-self.widths*(torch.exp(-r_ij)-self.centers)**2).cuda()
+        
+        return g_ij
+
+
+def _switch_component(x, ones, zeros):
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:226]: duplicated

Comment (ktschuett) [reply src/schnetpack/representation/support.py:226]: i think...

File: src/schnetpack/representation/support.py [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+    x_ = torch.where(x <= 0, ones, x)
+    return torch.where(x <= 0, zeros, torch.exp(-ones/x_))
+
+def switch_function(x, cuton, cutoff):
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:230]: duplicate

File: src/schnetpack/representation/support.py [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+    def forward(self, d_ij: torch.Tensor):
+
+
+        # Compute values of cutoff function
+        input_cut = 1 - 6*(d_ij/self.cutoff)**5 + 15*(d_ij/self.cutoff)**4 - 10*(d_ij/self.cutoff)**3
+        # Remove contributions beyond the cutoff radius
+        input_cut *= (d_ij < self.cutoff).float()
+        return input_cut
+    
+class SSP(nn.Module):
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:196]: this should not be needed...

File: src/schnetpack/representation/support.py [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+        vm = self.sequential_i(x)
+        v = snn.scatter_add(vp, idx_i, dim_size=n_atoms) + vm
+        for module in self.residual:
+            v = module(v)
+        
+        v = self.activation(v)
+        
+        return xp + self.linear_f(v)
+    
+class PhysNetCutOff(nn.Module):
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:181]: duplicate

File: src/schnetpack/representation/support.py [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+        )
+  
+    def forward(self,x):
+        
+        for module in self.residual:
+            x = module(x)
+        x = self.activation(x)
+        return x
+    
+class Module(nn.Module):
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:63]: change name. PhysNetModule?

File: src/schnetpack/representation/support.py [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+            self.linear,
+            activation(),
+            self.linear
+        )
+        
+    def forward(self,x):
+        return self.sequential(x) + x    
+    
+    
+class Output(nn.Module):
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:38]: rename. PhysNetOutput?

File: src/schnetpack/representation/support.py [hunk 19]
```diff
@@ -0,0 +1,237 @@
+import torch
+import torch.nn as nn
+import schnetpack.nn as snn
+from schnetpack.representation.weight_init import *
+
+from typing import Callable, Dict
+
+import schnetpack.properties as structure
+
+class Residual(nn.Module):
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:10]: rename. PhysNetResidual?

File: src/schnetpack/representation/support.py [hunk 20]
```diff
@@ -0,0 +1,237 @@
+import torch
+import torch.nn as nn
+import schnetpack.nn as snn
+from schnetpack.representation.weight_init import *
+
+from typing import Callable, Dict
+
+import schnetpack.properties as structure
+
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:9]: this can all be moved to the physnet package

File: src/schnetpack/representation/support.py [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+        n_interactions: int
+    ):
+
+        super(PhysNetInteraction, self).__init__()
+        self.n_atom_basis = n_atom_basis
+        self.n_rbf = n_rbf
+        self.activation = activation()
+        self.n_interactions = n_interactions
+        
+        self.linear_f = nn.Linear(n_atom_basis, n_atom_basis)
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:125]: use the schnetpack.nn.Dense, so you can directly give the activation.

File: src/schnetpack/representation/support.py [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+        self.linear_i.bias.data.fill_(0.)
+        self.linear_f.bias.data.fill_(0.)
+        
+        self.residual = nn.ModuleList(
+            [
+               Residual(n_atom_basis, activation) 
+               for _ in range(n_interactions)
+            ]
+        )
+        self.sequential_i = nn.Sequential(
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:147]: define this directly, so you don't need to list the layers separately above

File: src/schnetpack/representation/support.py [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+        x: torch.Tensor,
+        g_ij: torch.Tensor,
+        idx_i: torch.Tensor,
+        idx_j: torch.Tensor,
+        n_atoms: int
+    ):
+        
+        x_j = x[idx_j]
+        xp = self.u*x ## check this
+        vp = self.sequential_j(x_j.float())*self.linear_g(g_ij.float())
```
Comment (ktschuett) [inline src/schnetpack/representation/support.py:171]: no explicit casting

File: src/schnetpack/representation/weight_init.py [hunk 24]
```diff
@@ -0,0 +1,64 @@
+import torch
+import numpy as np
+
+#generates a random square orthogonal matrix of dimension dim
+def square_orthogonal_matrix(dim=3, seed=None):
```
Comment (ktschuett) [inline src/schnetpack/representation/weight_init.py:5]: this is a manual QR decomposition afaik. you can ust torch.qr directly and then use Q

File: src/schnetpack/representation/weight_init.py [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+        mat[n-1:, n-1:] = Hx
+        H = np.dot(H, mat)
+        # Fix the last sign such that the determinant is 1
+    D[-1] = (-1)**(1-(dim % 2))*D.prod()
+     # Equivalent to np.dot(np.diag(D), H) but faster, apparently
+    H = (D*H.T).T
+    return H
+
+#generates a random (semi-)orthogonal matrix of size NxM
+def semi_orthogonal_matrix(N, M, seed=None):
```
Comment (ktschuett) [inline src/schnetpack/representation/weight_init.py:27]: qr again

File: src/schnetpack/representation/weight_init.py [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+        square_matrix = square_orthogonal_matrix(dim=M, seed=seed)
+    return square_matrix[:N,:M]
+
+#generates a weight matrix with variance according to Glorot initialization
+#based on a random (semi-)orthogonal matrix
+#neural networks are expected to learn better when features are decorrelated
+#(stated by eg. "Reducing overfitting in deep networks by decorrelating representations",
+#"Dropout: a simple way to prevent neural networks from overfitting",
+#"Exact solutions to the nonlinear dynamics of learning in deep linear neural networks")
+def semi_orthogonal_glorot_weights(n_in, n_out, scale=2.0, seed=None):
```
Comment (ktschuett) [inline src/schnetpack/representation/weight_init.py:40]: there is actually `torch.nn.init.orthogonal_` now

File: src/schnetpack/representation/weight_init.py [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+    W = semi_orthogonal_matrix(n_in, n_out)
+    W *= np.sqrt(scale/((n_in+n_out)*W.var())) 
+    return W
+
+#given a tensor x, gives back the variance 
+def reduce_var(x, axis=None, keepdims=False):
```
Comment (ktschuett) [inline src/schnetpack/representation/weight_init.py:46]: https://pytorch.org/docs/stable/generated/torch.var.html?highlight=torch%20var#torch.var

File: test.ipynb [hunk 28]
```diff
@@ -0,0 +1,206 @@
+{
```
Comment (ktschuett) [inline test.ipynb:1]: should this be here?

**Review by ktschuett — CHANGES_REQUESTED** [review]

File: src/schnetpack/atomistic/D3.py [hunk 29]
```diff
@@ -0,0 +1,217 @@
...
+        self.k2 = k2 
+        self.k4 = k4
+        self.k5 = k5
+        self.k6 = k6
+        self.kn = kn
+        self.wf = wf
+        
+        #load D4 data 
+        directory = os.path.join(os.path.dirname(os.path.abspath(__file__)),'d4data')
+        self.register_buffer('refsys',torch.load(os.path.join(directory,'refsys.pth'))[:Zmax]) #[Zmax,max_nref]
```
[the remark this answers is not in the document]
Comment (MaxH1996) [reply src/schnetpack/atomistic/D3.py:87]: they come from the d4 folder in the atomistic folder

File: test.ipynb (hunk 28 above)
Comment (MaxH1996) [reply test.ipynb:1]: I will delete it in the final, for now it is just for testing.

**Pull request closed without merging**